lib/AST/ASTImporter.cpp

```diff
@@ -2591,14 +2616,30 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   if (ImportTemplateInformation(D, ToFunction))
     return nullptr;
 
+  bool IsFriend = D->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend);
+
   // Add this function to the lexical context.
   // NOTE: If the function is templated declaration, it should be not added into
   // LexicalDC. But described template is imported during import of
   // FunctionTemplateDecl (it happens later). So, we use source declaration
   // to determine if we should add the result function.
-  if (!D->getDescribedFunctionTemplate())
+  if (!D->getDescribedFunctionTemplate() && !IsFriend)
     LexicalDC->addDeclInternal(ToFunction);
 
+  // Friend declarations lexical context is the befriending class, but the
+  // semantic context is the enclosing scope of the befriending class.
+  // We want the friend functions to be found in the semantic context by lookup.
+  // FIXME should we handle this genrically in VisitFriendDecl?
+  // In Other cases when LexicalDC != DC we don't want it to be added,
+  // e.g out-of-class definitions like void B::f() {} .
+  if (LexicalDC != DC && IsFriend) {
```

Comment:
Alternative could be:
```
if (IsFriend) {
  assert(LexicalDC != DC);
  DC->makeDeclVisibleInContext(ToFunction);
}
```

Reply:
Ok, so friend declarations are missing from the enclosing class's declaration context (which is the LexicalDC) on purpose?

I wonder if we are importing only this:
```
struct X { friend void f(); };
```
Without any other declaration of `f`.

Should we still make f visible in DC? Or am I missing something?

Reply:
> Ok, so friend declarations are missing from the enclosing class's declaration context (which is the LexicalDC) on purpose?
Yes. Only the `FriendDecl` is part of the enclosing class's DC. The underlying `FunctionDecl` should not be part of the DC. But is should be available to lookup. 
Actually, it is one thing to be part of the list of decls of a DC; and an other thing is to be in the lookup table of a DC. By calling `makeDeclVisibleInContext` we put the decl only into the table but not into the list.

Reply:
```
struct X { friend void f(); };
```
In this case, yes we still make f visible to lookup, but it will not be part of the list of decls.

Reply:
My understanding was that DC is the DC where the class lives and LexicalDC is the class itself. So we want the `f` visible in DC but not in LexicalDC. But maybe my understanding was wrong.

Reply:
> DC is the DC where the class lives and LexicalDC is the class itself. 
Right.

> So we want the f visible in LexicalDC but not in DC
No, we want f visible in DC and not visible in LexicalDC.

Reply:
We had a discussion about whether calling `makeDeclVisibleInContext` would break the original semantics in case of this code:
```
struct X { friend void f(); }; void g() { f(); } // Error, f is not visible for normal lookup
```

And my answer is it will not.The imported AST will have the same semantics. Because the way normal C++ lookup works (implemented in `clang::Sema::CppLookupName()` and in `LookupDirect()`) is that it will return the found `NamedDecl` only if the set IDNS is matching:
```
  // Perform lookup into this declaration context.
  DeclContext::lookup_result DR = DC->lookup(R.getLookupName());
  for (NamedDecl *D : DR) {
    if ((D = R.getAcceptableDecl(D))) { // HERE we check the IDNS
      R.addDecl(D);
      Found = true;
    }
  }
```
During the parsing of the body of `g()` a lookup will be initiated but with IDNS_Ordinary.

Added some more tests and updated the PR accordingly.